### ISSUE TITLE
Restructure `multi_head_attention_forward`

### ISSUE THREAD

**Issue by Enealor** [pytorch/pytorch#34573]
## 🚀 Feature

Restructure the function `multi_head_attention_forward`  in [nn.functional](https://github.com/pytorch/pytorch/blob/23b2fba79a6d2baadbb528b58ce6adb0ea929976/torch/nn/functional.py#L3573) into several functions to improve the ability to experiment. In particular, decompose the function so that the following are available:
* The input embedding functions.
* The computation of attention weights.
* The output embedding function.

This will allow users to try different embeddings or attention mechanisms without having to recode the rest.

## Motivation

Addresses the issue of decomposing the function as mentioned in #32590. It also moves forward on including more support for attention mechanisms.

## Pitch

Currently, the `mutli_head_attention_forward` function encapsulates the projection of the `query`, `key`, and `value`, computing attention for these projections, and computing the output projection after applying attention. Furthermore, the input embedding utilizes several code paths that are different embeddings. By decomposing the function into several parts, we can make it more readable and open to experimentation.

The following breakdown is based on the above:

* Functions for computing the input embeddings `q`, `k`, and `v`. There are currently four code paths used for doing this, and each should be an individual function so that it's clearer what is the intended outcome. (See L3669-L3748.)
* A function for computing the attention weight. Some models rely on computing attention in different ways, and separating this out would allow us to use those more freely. (See L3750-L3824).
* A function for computing the output projection from the attention weight and `value` projection. (See L3826-L3836.)

**Comment by ngimel** [comment]
cc @zhangguanheng66 @cpuhrsch.
@enealor do you want to work on this? If so, can you please coordinate with @zhangguanheng66 and @cpuhrsch . 
Thanks!

**Comment by Enealor** [comment]
@ngimel I am happy to coordinate with them. The issue that spawned this asked for helpers to first open an issue and wait for feedback. I am willing to work on it further given the go-ahead.

**Comment by cpuhrsch** [comment]
@Enealor - We're actively working on this for several of the reasons you stated. We'll need review / input of our designs and eventually also work to implement. I'm happy to see we have community support for this and are happy to loop you in.

**Comment by zhangguanheng66** [comment]
Definitely. As the first step, please split the MHA function into several building blocks in nn.Module and nn.functional. Something like [this](https://github.com/zhangguanheng66/torchBERT/blob/6becd06ee049b9e0d35dc222b2c3410473263664/torchBERT/model.py#L45-L105) should be a good start. Feel free to pin us once you have a PR. 

Please keep in mind that those building blocks should be JIT-ability, quantization, and potentially porting to C++ in the future.

**Comment by zhangguanheng66** [comment]
@Enealor I starts a PR [here](https://github.com/pytorch/pytorch/pull/34719). Feel free to continue the work if you have time. After we collect some feedback, we can move the code to `torch.nn.functional`

**Comment by Enealor** [comment]
@zhangguanheng66 Thank you for the link. I have started restructuring the code in `functional`. Based on my work so far, I have changed my pitch and added an alternative.

I could use feedback on how to handle the input embeddings. As I explain above, there are three different ways the input embedding is calculated. Should I add a function for each, or provide one function that can be used by the end-user to build up the rest? I ask because I had started the former, while I noticed in the PR that you are doing the latter.

**Comment by Enealor** [comment]
Follow-up: I have created a PR with the approach that would introduce less specialized functions. I believe it is ready for feedback.

**Comment by zhangguanheng66** [comment]
> Follow-up: I have created a PR with the approach that would introduce less specialized functions. I believe it is ready for feedback.

Thanks. @Enealor Made some comments there. Just wondering if you want to merge your work to this one. Then, you can take charge of this PR from now on.

**Comment by netw0rkf10w** [comment]
Hi guys! 
Could you please tell me what is the current state of this?
I would like to implement a different method for computing the attention weights, and was about to split the current `multi_head_attention_forward` into smaller functions, but then fortunately I found this thread so I thought maybe I shouldn't start from scratch.
Thanks a lot!

**Comment by zhangguanheng66** [comment]
The new MHA has been landed in torchtext. 
https://github.com/pytorch/text/blob/60907bf3394a97eb45056a237ca0d647a6e03216/torchtext/modules/multiheadattention.py#L5

**Comment by netw0rkf10w** [comment]
@zhangguanheng66 Awesome! Thanks a lot!
The code looks very clean. Is there a plan to bring this to the core torch?

I observed the use of `torch.jit.is_scripting()` in the current implementation in `torch.nn.functional` ([here](https://github.com/pytorch/pytorch/blob/5e03a1e9261e86d3ee62481ef40d98ca5f09646c/torch/nn/functional.py#L3915)), but on in your `torchtext` implementation. Would you have some comments on why there's a difference on the relative performance of the two? Thanks.

**Comment by zhangguanheng66** [comment]
> @zhangguanheng66 Awesome! Thanks a lot!
> The code looks very clean. Is there a plan to bring this to the core torch?
> 
> I observed the use of `torch.jit.is_scripting()` in the current implementation in `torch.nn.functional` ([here](https://github.com/pytorch/pytorch/blob/5e03a1e9261e86d3ee62481ef40d98ca5f09646c/torch/nn/functional.py#L3915)), but on in your `torchtext` implementation. Would you have some comments on why there's a difference on the relative performance of the two? Thanks.

We have some old usage cases that need it. The new implementation doesn't require this for torchscript support. Let us know if you have any feedback (We plan to include it in torchtext 0.7.0 release by the end of July). For now, we don't have a plan to move it to pytorch core library because we don't want to have two MHA modules in the same library. Plus, the MHA container is more research oriented so it's better to stay in torchtext.

**Comment by netw0rkf10w** [comment]
@zhangguanheng66 Thanks for the reply.

What I plan to do is to refactor `torch.nn.Transformer` to use your new containers, with backward compatibility: loading weights trained with the "previous" implementation must work seamlessly. Do you think this is possible? If yes then moving your code to the core won't break anything...

**Comment by zhangguanheng66** [comment]
> @zhangguanheng66 Thanks for the reply.
> 
> What I plan to do is to refactor `torch.nn.Transformer` to use your new containers, with backward compatibility: loading weights trained with the "previous" implementation must work seamlessly. Do you think this is possible? If yes then moving your code to the core won't break anything...

Currently, we don't have a plan to support the pre-trained transformer model with the new MHA container. It was developed to grant flexibility to future research work.

**Comment by netw0rkf10w** [comment]
@zhangguanheng66 I am working on making your MHA container a drop-in replacement of PyTorch's MHA. The first result would be having all `nn.MultiheadAttention` instances in `nn.Transformer` replaced with the MHA container. I am close, but there's still something that needs to be taken into account: the [`key_padding_mask`](https://github.com/pytorch/pytorch/blob/67a4f375cdf06113ca959b4e16739edb666f243f/torch/nn/modules/activation.py#L904) argument. It's still unclear to me what this argument does exactly.

I can see that you've recently attempted to rebuild BERT using the MHA container. However, you did not use `key_padding_mask` despite [it being declared](https://github.com/pytorch/text/blob/9ddcb829752fd37b1c343c932e77f6899408722c/examples/BERT/model.py#L92). Could you please explain? Thanks.

**Comment by zhangguanheng66** [comment]
> @zhangguanheng66 I am working on making your MHA container a drop-in replacement of PyTorch's MHA. The first result would be having all `nn.MultiheadAttention` instances in `nn.Transformer` replaced with the MHA container. I am close, but there's still something that needs to be taken into account: the [`key_padding_mask`](https://github.com/pytorch/pytorch/blob/67a4f375cdf06113ca959b4e16739edb666f243f/torch/nn/modules/activation.py#L904) argument. It's still unclear to me what this argument does exactly.
> 
> I can see that you've recently attempted to rebuild BERT using the MHA container. However, you did not use `key_padding_mask` despite [it being declared](https://github.com/pytorch/text/blob/9ddcb829752fd37b1c343c932e77f6899408722c/examples/BERT/model.py#L92). Could you please explain? Thanks.

In the BERT example, we don't use `key_padding_mask` because tokens in the sequence are masked in the embedding layer. Since we provide a 3-D tensor for the `attn_mask`, if you want to mask a padding token, you can set all the corresponding numbers in `attn_mask` to `True` so no attention is generated for the padding token.

**Comment by netw0rkf10w** [comment]
> Since we provide a 3-D tensor for the `attn_mask`, if you want to mask a padding token, you can set all the corresponding numbers in `attn_mask` to `True` so no attention is generated for the padding token.

Yes this is what I was thinking as well, but then why is `key_padding_mask` needed in `functional.multi_head_attention_forward()`? I'm confused...

**Comment by zhangguanheng66** [comment]
> > Since we provide a 3-D tensor for the `attn_mask`, if you want to mask a padding token, you can set all the corresponding numbers in `attn_mask` to `True` so no attention is generated for the padding token.
> 
> Yes this is what I was thinking as well, but then why is `key_padding_mask` needed in `functional.multi_head_attention_forward()`? I'm confused...

That's a decision we made more than 1 year ago. And once merged in the release, it's hard to remove or change for BC breaking reason. For the new implementation here (a.k.a. MHA container), I try to simplify the APIs.

**Comment by netw0rkf10w** [comment]
@zhangguanheng66 I see. Yes the new implementation is very clean, thanks for that.

> Since we provide a 3-D tensor for the `attn_mask`, if you want to mask a padding token, you can set all the corresponding numbers in `attn_mask` to `True` so no attention is generated for the padding token.

Let me make this more concrete. In the old implementation we have:

```
key_padding_mask: (N, S)
attn_mask: (L, S) for 2D mask and (N*H, L, S) for 3D mask
```

where `L`: target length, `S`: source length, `N`: batch size, `H`: number of heads. In the new one, we only have:

`attn_mask: (N * H, L, S) `

Now if we are given `key_padding_mask` and we want to use the new MHA, we need to merge the masks:

`attn_mask = attn_mask | key_padding_mask.unsqueeze(1).repeat(H, L, 1)`

Is that correct?

By the way, are you planning to re-implement the Transformer based on the new MHA?
Thanks.

**Comment by netw0rkf10w** [comment]
> In the BERT example, we don't use `key_padding_mask` because tokens in the sequence are masked before the embedding layer.

Wait, why did you add `src_key_padding_mask` as an argument then?

**Comment by netw0rkf10w** [comment]
> Let me make this more concrete. In the old implementation we have:
> 
> ```
> key_padding_mask: (N, S)
> attn_mask: (L, S) for 2D mask and (N*H, L, S) for 3D mask
> ```
> 
> where `L`: target length, `S`: source length, `N`: batch size, `H`: number of heads. In the new one, we only have:
> 
> `attn_mask: (N * H, L, S) `
> 
> Now if we are given `key_padding_mask` and we want to use the new MHA, we need to merge the masks:
> 
> `attn_mask = attn_mask | key_padding_mask.unsqueeze(1).repeat(H, L, 1)`
> 
> Is that correct?
> 
> By the way, are you planning to re-implement the Transformer based on the new MHA?
> Thanks.

The above is not good. To convert `key_padding_mask` to a 3D mask, e.g. for self-attention, one should do:

`key_padding_mask_3D = ~torch.einsum('bi,bj->bij', ~key_padding_mask, ~key_padding_mask).repeat(H, 1, 1)`

**Comment by zhangguanheng66** [comment]
> @zhangguanheng66 I see. Yes the new implementation is very clean, thanks for that.
> 
> > Since we provide a 3-D tensor for the `attn_mask`, if you want to mask a padding token, you can set all the corresponding numbers in `attn_mask` to `True` so no attention is generated for the padding token.
> 
> Let me make this more concrete. In the old implementation we have:
> 
> ```
> key_padding_mask: (N, S)
> attn_mask: (L, S) for 2D mask and (N*H, L, S) for 3D mask
> ```
> 
> where `L`: target length, `S`: source length, `N`: batch size, `H`: number of heads. In the new one, we only have:
> 
> `attn_mask: (N * H, L, S) `
> 
> Now if we are given `key_padding_mask` and we want to use the new MHA, we need to merge the masks:
> 
> `attn_mask = attn_mask | key_padding_mask.unsqueeze(1).repeat(H, L, 1)`
> 
> Is that correct?
> 
> By the way, are you planning to re-implement the Transformer based on the new MHA?
> Thanks.

Yes. You are right to combine `attn_mask` and `key_padding_mask`. Let me know if it works.

For the second question, the answer is no. For the transformer and MHA in the core library, we want to minimize the changes unless there is a significant bug. This is to avoid BC breaking considering the large user group.

**Comment by netw0rkf10w** [comment]
> For the second question, the answer is no. For the transformer and MHA in the core library, we want to minimize the changes unless there is a significant bug. This is to avoid BC breaking considering the large user group.

Sorry my question was about `torchtext` (not the core PyTorch). You've switched to the new MHA in the BERT example, so I was wondering if you would add Transformer or not. I think `torchtext` should also has its own version of the Transformer.

**Comment by zhangguanheng66** [comment]
> > For the second question, the answer is no. For the transformer and MHA in the core library, we want to minimize the changes unless there is a significant bug. This is to avoid BC breaking considering the large user group.
> 
> Sorry my question was about `torchtext` (not the core PyTorch). You've switched to the new MHA in the BERT example, so I was wondering if you would add Transformer or not. I think `torchtext` should also has its own version of the Transformer.

what benefits do you think if we add a new version of Transformer, with the only difference of MHA layer? If you like, you can create a custom `TransformerEncoderLayer` with the new MHA container and throw it to `torch.nn.Transformer`.

**Comment by netw0rkf10w** [comment]
@zhangguanheng66 Actually I already have my own version of the Transformer based on your MHA container. All the old MHA layers (in both the encoder and decoder) are replaced by the new one, and as a consequence, all the `*_key_padding_mask` arguments are removed.

The reason I think `torchtext` should have a new version of the Transformer documented (even as an example just like the BERT one) is to encourage people to start using the new MHA implementation. My vision for the future is that this should become the default implementation in the core PyTorch (because the current one is really a mess, sorry to say that). Currently this is not possible because of the large user group, but when most of the users use the new implementation, why not ;)

**Comment by zhangguanheng66** [comment]
> @zhangguanheng66 Actually I already have my own version of the Transformer based on your MHA container. All the old MHA layers (in both the encoder and decoder) are replaced by the new one, and as a consequence, all the `*_key_padding_mask` arguments are removed.
> 
> The reason I think `torchtext` should have a new version of the Transformer documented (even as an example just like the BERT one) is to encourage people to start using the new MHA implementation. My vision for the future is that this should become the default implementation in the core PyTorch (because the current one is really a mess, sorry to say that). Currently this is not possible because of the large user group, but when most of the users use the new implementation, why not ;)

Do you think the `TransformerEncoderLayer` and `TransformerEncoder` [here](https://github.com/pytorch/text/blob/49c558b2d3e39dbd7d7bb9a70ce89e663a648903/examples/BERT/model.py#L52-L113) could be a good example?

Actually, I'm interested to see your work to combine `attn_mask` and `key_padding_mask` in MHA container. Do you want to create a PR for that?

If you propose some work in torchtext, please don't hesitate to open an issue [here](https://github.com/pytorch/text/issues).

**Comment by netw0rkf10w** [comment]
> Do you think the `TransformerEncoderLayer` and `TransformerEncoder` [here](https://github.com/pytorch/text/blob/49c558b2d3e39dbd7d7bb9a70ce89e663a648903/examples/BERT/model.py#L52-L113) could be a good example?

Definitely! In my implementation I defined the following layer `MultiheadAttention` that closely matches `nn.MultiheadAttention` (in terms of function arguments, for ease of replacement):

```
class MultiheadAttention(torch.nn.Module):
    def __init__(self, embed_dim, nhead, dropout=0.1, bias=True, add_bias_kv=False,
                 kdim=None, vdim=None):
        super().__init__()
        kdim = embed_dim if kdim is None else kdim
        vdim = embed_dim if vdim is None else vdim

        in_proj_container = InProjContainer(torch.nn.Linear(embed_dim, embed_dim, bias=bias),
                                        torch.nn.Linear(kdim, embed_dim, bias=bias),
                                        torch.nn.Linear(vdim, embed_dim, bias=bias))
        self.mha = MultiheadAttentionContainer(nhead,
                                        in_proj_container,
                                        ScaledDotProduct(dropout=dropout),
                                        torch.nn.Linear(embed_dim, embed_dim))
        if add_bias_kv:
            self.bias_k = torch.nn.Parameter(torch.empty(1, 1, embed_dim))
            self.bias_v = torch.nn.Parameter(torch.empty(1, 1, embed_dim))
        else:
            self.bias_k = self.bias_v = None


    def forward(self, query, key, value, attn_mask=None):
        return self.mha(query, key, value, attn_mask=attn_mask, bias_k=self.bias_k,
                         bias_v=self.bias_v)
```

> 
> Actually, I'm interested to see your work to combine `attn_mask` and `key_padding_mask` in MHA container. Do you want to create a PR for that?
> 
> If you propose some work in torchtext, please don't hesitate to open an issue [here](https://github.com/pytorch/text/issues).

Sure, I'll create a PR in the next few days (and then let's continue the discussion in the `torchtext` repo). Thanks!

**Comment by JaireYu** [comment]
Hi, everyone. I have a stupid question when I read the multi_head_attention_forward function. 
```py
 if static_k is None and static_v is None:
            k = torch.cat([k, bias_k.repeat(1, bsz, 1)])
            v = torch.cat([v, bias_v.repeat(1, bsz, 1)])
            if attn_mask is not None:
                attn_mask = pad(attn_mask, (0, 1))
            if key_padding_mask is not None:
                key_padding_mask = pad(key_padding_mask, (0, 1)) 
```
why cat the bias to the sequence, not to add on the sequence. What is it used for? It there a reference paper clarifying this?

**Comment by zhangguanheng66** [comment]
> Hi, everyone. I have a stupid question when I read the multi_head_attention_forward function.
> 
> ```python
>  if static_k is None and static_v is None:
>             k = torch.cat([k, bias_k.repeat(1, bsz, 1)])
>             v = torch.cat([v, bias_v.repeat(1, bsz, 1)])
>             if attn_mask is not None:
>                 attn_mask = pad(attn_mask, (0, 1))
>             if key_padding_mask is not None:
>                 key_padding_mask = pad(key_padding_mask, (0, 1)) 
> ```
> 
> why cat the bias to the sequence, not to add on the sequence. What is it used for? It there a reference paper clarifying this?

It is used when key/value are statics, and the function attaches one more token on each iteration to key/value.

**Comment by JaireYu** [comment]
> > Hi, everyone. I have a stupid question when I read the multi_head_attention_forward function.
> > ```python
> >  if static_k is None and static_v is None:
> >             k = torch.cat([k, bias_k.repeat(1, bsz, 1)])
> >             v = torch.cat([v, bias_v.repeat(1, bsz, 1)])
> >             if attn_mask is not None:
> >                 attn_mask = pad(attn_mask, (0, 1))
> >             if key_padding_mask is not None:
> >                 key_padding_mask = pad(key_padding_mask, (0, 1)) 
> > ```
> > 
> > 
> > why cat the bias to the sequence, not to add on the sequence. What is it used for? It there a reference paper clarifying this?
> 
> It is used when key/value are statics, and the function attaches one more token on each iteration to key/value.

Thank  you for your reply. But  I know on which condition it  is needed. I don't know why we  need to attach the additional token.

**Comment by zhangguanheng66** [comment]
This is used when previous time steps are cached - no need to recompute key and value (since they are static here). However, for some tasks (language modeling), this attaches a new token to key/value even though they are static.
